specutils/manipulation/resample.py

```diff
@@ -0,0 +1,293 @@
+        if orig_spectrum.uncertainty is not None:
+            if isinstance(orig_spectrum.uncertainty, StdDevUncertainty):
+                pixel_uncer = np.square(orig_spectrum.uncertainty.array)
+            elif isinstance(orig_spectrum.uncertainty, VarianceUncertainty):
+                pixel_uncer = orig_spectrum.uncertainty.array
+            elif isinstance(orig_spectrum.uncertainty, InverseVariance):
+                pixel_uncer = np.reciprocal(orig_spectrum.uncertainty.array)
+        else:
+            pixel_uncer = None
+
+        # todo: Current code doesn't like the inputs being quantity objects, may
+        # want to look into this more in the future
+        resample_grid = self._resample_matrix(np.array(orig_spectrum.wavelength),
+                                              np.array(fin_lamb))
+
+        # Now for some broadcasting magic to handle multi dimensional flux inputs
+        # Essentially this part is inserting length one dimensions as fillers
+        # For example, if we have a (5,6,10) input flux, and an output grid
+        # of 3, flux will be broadcast to (5,6,1,10) and resample_grid will
+        # Be broadcast to (1,1,3,10).  The sum then reduces down the 10, the
+        # original dispersion grid, leaving 3, the new dispersion grid, as
+        # the last index.
+        new_flux_shape = list(orig_spectrum.flux.shape)
+        new_flux_shape.insert(-1, 1)
+        in_flux = orig_spectrum.flux.reshape(new_flux_shape)
+
+        ones = [1] * len(orig_spectrum.flux.shape[:-1])
+        new_shape_resample_grid = ones + list(resample_grid.shape)
+        resample_grid = resample_grid.reshape(new_shape_resample_grid)
+
+        # Calculate final flux
+        out_flux = np.sum(in_flux * resample_grid, axis=-1) / np.sum(
+            resample_grid, axis=-1)
+
+        # Calculate output uncertainty
+        if pixel_uncer is not None:
+            pixel_uncer = pixel_uncer.reshape(new_flux_shape)
+
+            out_variance = np.sum(pixel_uncer * resample_grid**2, axis=-1) / np.sum(
+                resample_grid**2, axis=-1)
+            out_uncertainty = InverseVariance(np.reciprocal(out_variance))
+        else:
+            out_uncertainty = None
+
+        # todo: for now, use the units from the pre-resampled
+        # spectra, although if a unit is defined for fin_lamb and it doesn't
+        # match the input spectrum it won't work right, will have to think
+        # more about how to handle that... could convert before and after
+        # calculation, which is probably easiest. Matrix math algorithm is
+        # geometry based, so won't work to just let quantity math handle it.
+        resampled_spectrum = Spectrum1D(np.nan_to_num(out_flux),
+                                        np.array(fin_lamb) * orig_spectrum.wavelength.unit,
+                                        uncertainty=np.nan_to_num(out_uncertainty))
+
+        return resampled_spectrum
+
+
+class InterpolatedResample(ResampleBase):
```

Comment:
I think this is better just split into two separate classes - `LinearInterpolatedResample` and `SplineInterpolatedResample`.  I'm thinking this because the `SplineInterpolatedResample` should get some keywords to set things like the order of the spline, but those don't apply in the linear case, so instead of having some confusing rule like "this keyword is ignored when interp_type is linear but not spline" you can just do separate classes.